lib/learnosity/sdk/request/init.rb

```diff
@@ -3,226 +3,271 @@
-	      raise ValidationException, 'Questions API requires a `user_id` in the security packet'
-	    end
+          unless @action.nil?
+            signature_array << @action
+          end
 
-	    if ! @security_packet.include? 'timestamp'
-	      @security_packet['timestamp'] = Time.now.gmtime.strftime('%Y%m%d-%H%m')
-	    end
-	  end
+          hash_signature(signature_array)
+        end
 
-	  if @secret.nil? or ! @secret.is_a? String
-	   raise ValidationException, 'The `secret` argument must be a valid string'
-	  end
+        def generate(encode = true)
+          output = {}
+
+          case @service
+          when 'assess', 'author', 'data', 'items', 'reports'
+            output['security'] =  @security_packet
+
+            unless @request_packet.nil?
+              output['request'] = @request_packet
+            end
+
+            case @service
+            when 'data'
+              data_output = { 'security' => JSON.generate(output['security']) }
+              if output.key?('request')
+                data_output['request'] = JSON.generate(output['request'])
+              end
+              unless @action.nil?
+                data_output['action'] = @action
+              end
+              return data_output
+
+            when 'assess'
+              output = output['request']
+            end
+
+          when 'questions'
+            output = hash_except(@security_packet, 'domain')
+            unless @request_packet.nil?
+              output = output.merge(@request_packet)
+            end
+
+          when 'events'
+            output['security'] =  @security_packet
+            output['config'] =  @request_packet
+          else
+            raise Exception, "generate() for #{@service} not implemented"
+          end
+
+          unless encode
+            return output
+          end
+          JSON.generate(output)
+        end
 
-	  # XXX we don't do JSON to native object conversion for now, as the PHP SDK does
+        protected
+
+        attr_accessor :service, :secret, :request_packet, :action, :sign_request_data
+        attr_writer :security_packet, :request_string
+
+        def add_meta
+          if @request_packet.nil?
+            @request_packet = {}
+          end
+
+          if Sys::Platform.linux?
+            platform = 'linux'
+          elsif Sys::Platform.windows?
+            platform = 'win'
+          elsif Sys::Platform.mac?
+            platform = 'darwin'
+          else
+            platform = Sys::Uname.platform
+          end
+
+          sdk_metrics = {
+            :version => VERSION,
+            :lang => 'ruby',
+            :lang_version => RUBY_VERSION,
+            :platform => platform,
+            :platform_version => Sys::Uname.release
+          }
+
+          if @request_packet.include? 'meta'
+            @request_packet['meta'][:sdk] = sdk_metrics
+          elsif @request_packet.include? :meta
+            @request_packet[:meta][:sdk] = sdk_metrics
```

Comment:
I mentioned this in another comment but is there actually a difference between the symbol and the string version of this? 😱

Reply:
Unfortunately yes. We can't guarantee what type of input the user will provide so best to account for both symbols and strings as keys... fwiw, symbols is the more 'rubyesque' method.